### PR TITLE
[WIP] microlens-pro

### PR DESCRIPTION
**Pull request by crumbtoo**
Draft PR to track progress on `microlens-pro`.
Original issue: #105 
Original WIP PR: #107 (which will serve as reference if not the base of this)

Are there any objections to following the path described in the original thread; that is, re-exporting `microlens-platform` with the imposter Prisms hidden and replaced with the lens-accurate definitions?

### CODE REVIEW

**Comment by crumbtoo** [comment]
Last commit is a nice starting point for discussion. Included are at least replacements for the faux Prisms and Isos from `microlens`, as well as a number of other iso/prism-related definitions.

The main points of concern from the original thread included handling of `non`, `_Just`, `_Left`, etc. from `Lens.Micro`. Currently, they are redefined as proper Isos and Prisms in `Lens.Micro.Pro`. An accompanying module `Lens.Micro.ProCompat` re-exports the entirety of `Lens.Micro` with the exception of `_Just` and friends, which are sourced from `Lens.Micro`.

This scheme of wrapping `Lens.Micro` (or potentially even `microlens-platform`) was proposed by neongreen in the original issue, and alternate solutions are very welcome.

Another point of note is the obvious lack of documentation. microlens' documentation is leagues more approachable than lens', and thus should be handled with more care; i.e. not cut-and-pasted. a loose to-do list includes Template Haskell, and Haddocks.

**Comment by stevenfontanella** [comment]
Thanks for opening! Looks like you have good progress already.

> Are there any objections to following the path described in the original thread; that is, re-exporting microlens-platform with the imposter Prisms hidden and replaced with the lens-accurate definitions?

No objections, that sounds good to me. It sounds like the idea is that users would either import Lens.Micro / Lens.Micro.Platform or Lens.Micro.Pro depending on their situation. I think that's good.

Re: the rest of the package: I'm less familiar with these lens features. Can you help explain your thought process for what definitions to include and not include? I see that most come from [Control.Lens.Combinators](https://hackage.haskell.org/package/lens-5.2.3/docs/Control-Lens-Combinators.html) but naturally not everything. From #105 I understand that we want prisms and isomorphisms but it looks like there are several related definitions as well. Basically, I'm wondering how we should choose whether to include a definition here or not? That will help me review this.

File: microlens-pro/src/Lens/Micro/Pro/Internal.hs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        Right s -> case seta s of
+            Left t -> Left (Right t)
+            Right a -> Right a
+
+    {-# INLINE right' #-}
+    {-# INLINE left' #-}
+
+type Market' a s = Market a a s s
+
+-- TODO: IsText is up for discussion
```
Comment (stevenfontanella) [inline microlens-pro/src/Lens/Micro/Pro/Internal.hs:79]: Is the concern here that it would incur a dependency on `text`?

Comment (crumbtoo) [reply microlens-pro/src/Lens/Micro/Pro/Internal.hs:79]: Yeah. Not an issue if we're depending on microlens-platform I suppose

**Comment by crumbtoo** [comment]
I think all that I'm confident doing is done. I didn't update the README:

* Build times - I don't know what machine these measurements were taken on, nor the methods used to take them.
* There's a significant amount of text that explicitly discusses the absence of prisms and isomorphisms, and I'm not sure how to appropriately go about updating it.

**Comment by stevenfontanella** [comment]
Sounds good, I’ll take a look tonight. 

> Build times

I can measure these again for everything, it’s probably best to get more up to date numbers anyway. 

> updating the README

I can do this separately, no need to wait for that.

**Comment by stevenfontanella** [comment]
I rebased on master and force-pushed. If you make any additional changes, please just back them up to a separate branch, force-pull (please be careful not to lose any work that wasn't pushed here) and then cherry-pick again before pushing here. 

I updated the CI to include microlens-pro, let's try to get it working before merging. I see failures now that I'll take a look at.

File: microlens-pro/microlens-pro.cabal [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  which inspired documentation, and Wikibooks contributors for
+  <https://en.wikibooks.org/wiki/Haskell/Lenses_and_functional_references#Isos this article>.
+  .
+  This package is a part of the
+  <https://hackage.haskell.org/package/microlens microlens> family; see the
+  readme <https://github.com/stevenfontanell/microlens on Github>.
+license:                BSD3
+license-file:           LICENSE
+author:                 Edward Kmett, Artyom Kazak, crumbtoo <lomiskiam@gmail.com>
+maintainer:             Steven Fontanella <steven.fontanella@gmail.com>
```
Comment (stevenfontanella) [inline microlens-pro/microlens-pro.cabal:22]: I updated the author, maintainer and github links, please let me know if you see any issues with it. I changed it so that you're an author and I'm the maintainer, since it lives in this repo.

Comment (crumbtoo) [reply microlens-pro/microlens-pro.cabal:22]: Looks good!

**Comment by crumbtoo** [comment]
overly-strict cabal version bounds might be the issue with the CI. They were generated with cabal gen-bounds under GHC 9.6.4 if I remember correctly

**Comment by crumbtoo** [comment]
Couldn't get working on ghc-7.10 :(. Main obstacle is `GHC.Exts.TYPE`, which is seemingly not exported despite [docs](https://hackage.haskell.org/package/base-4.19.0.0/docs/GHC-Exts.html#t:TYPE) suggesting it has been since at least base-2.1...

File: microlens-pro/src/Lens/Micro/Pro/TH.hs [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+
+#if __GLASGOW_HASKELL__ >= 800
+{-# LANGUAGE TemplateHaskellQuotes #-}
+#else
+{-# LANGUAGE TemplateHaskell #-}
+#endif
+
+#ifndef MIN_VERSION_template_haskell
+#define MIN_VERSION_template_haskell(x,y,z) 1
+#endif
```
Comment (stevenfontanella) [inline microlens-pro/src/Lens/Micro/Pro/TH.hs:16]: What is this for? I don't see it referenced anywhere

File: microlens-pro/microlens-pro.cabal [hunk 4]
```diff
@@ -7,8 +7,8 @@ description:
   .
   Thank you to the <https://hackage.haskell.org/package/lens lens> contributors
   for the original code and some docs,
-  <https://cohomolo.gy Emily Pillmore> and
-  <https://mroman42.github.io/cosmoi Mario Román> for
+  Emily Pillmore (<https://cohomolo.gy link>) and
+  Mario Rom&#225;n (<https://mroman42.github.io/cosmoi link>) for
```
Comment (stevenfontanella) [inline microlens-pro/microlens-pro.cabal:11]: HTML escaping doesn't seem to work inside links in Hackage, so I pulled out the link

**Comment by stevenfontanella** [comment]
I left some non-blocking comments. I'll merge and publish the v1 when the CI passes. Thanks a lot for this big contribution!

File: microlens-pro/src/Lens/Micro/Pro/TH.hs (hunk 3 above)
Comment (crumbtoo) [reply microlens-pro/src/Lens/Micro/Pro/TH.hs:16]: Not sure. The entire TH module was cut-and-pasted from #107. Probably safe to remove...?

File: microlens-pro/src/Lens/Micro/ProCompat.hs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{- |
+Module      : Lens.Micro.ProCompat
+Copyright   : (C) 2013-2016 Edward Kmett, 2018 Monadfix
+License     : BSD-style (see the file LICENSE)
+
+This module re-exports "Lens.Micro.Platform", overriding definitions where
+necesarry.
+-}
+module Lens.Micro.ProCompat
```
[the remark this answers is not in the document]
Comment (crumbtoo) [reply microlens-pro/src/Lens/Micro/ProCompat.hs:9]: It looks like it's still exported from `Lens.Micro.Platform` and thus `Lens.Micro.ProCompat`, no?
```
ghci> import Lens.Micro.ProCompat
ghci> :t Lens.Micro.ProCompat.preview 
Lens.Micro.ProCompat.preview
  :: MonadReader s m =>
     Getting (Data.Monoid.First a) s a -> m (Maybe a)
```

File: microlens-pro/src/Lens/Micro/Pro/TH.hs (hunk 3 above)
Comment (stevenfontanella) [reply microlens-pro/src/Lens/Micro/Pro/TH.hs:16]: I see it comes from [here](https://github.com/ekmett/lens/blob/master/include/lens-common.h) and is used in a [couple places](https://github.com/search?q=repo%3Aekmett%2Flens%20MIN_VERSION_template_haskell&type=code) in the lens repo (but not here). Let me try to remove it.

File: microlens-pro/src/Lens/Micro/ProCompat.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (stevenfontanella) [reply microlens-pro/src/Lens/Micro/ProCompat.hs:9]: Got it, I think I tested it wrong. Looks good then!

**Comment by stevenfontanella** [comment]
I tried using the library a little bit. @crumbtoo, what do you think about changing `Lens.Micro.ProCompat` into `Lens.Micro.Pro`? i.e. perform the same exports directly in `Lens.Micro.Pro` and remove `Lens.Micro.ProCompat`? My thinking is that `Lens.Micro.Pro` is currently incompatible with both `Lens.Micro` and `Lens.Micro.Platform`, and when using `Lens.Micro.Pro`, you would likely end up needing definitions defined in `Lens.Micro` / `Lens.Micro.Platform`. 

Basically I don't see a reason to import `Lens.Micro.Pro` instead of `Lens.Micro.ProCompat`. In the former case, you would likely need to import `Lens.Micro` / `Lens.Micro.Platform` and add several `hiding` clauses. So I feel we can just only define `Lens.Micro.Pro` and let it do the same re-exports. Does that sound ok to you, or is there a reason not to do this?

**Comment by crumbtoo** [comment]
That's true :p. I don't have any objections.
